File: Sources/Lib/Processors/CompressorProcessor.swift
import Accelerate
import DSPAudio
import DSPConfig
import DSPFilters
import Foundation

final class CompressorProcessor: Processor {
  let name: String
  private var monitorChannels: [Int]
  private var processChannels: [Int]
  private var attack: PrcFmt
  private var release: PrcFmt
  private var threshold: PrcFmt
  private var factor: PrcFmt
  private var makeupGain: PrcFmt
  private var limiter: LimiterFilter?
  private var scratch: [PrcFmt]
  private var prevLoudness: PrcFmt = -100.0

  init(
    name: String = "compressor", parameters: CompressorParameters, sampleRate: Int, chunkSize: Int
  ) {
    self.name = name
    self.scratch = [PrcFmt](repeating: 0.0, count: chunkSize)

    var monitor = parameters.monitorChannelsArray()
    if monitor.isEmpty {
      monitor = Array(0..<parameters.channels)
    }
    self.monitorChannels = monitor

    var process = parameters.processChannelsArray()
    if process.isEmpty {
      process = Array(0..<parameters.channels)
    }
    self.processChannels = process

    let srate = PrcFmt(sampleRate)
    self.attack = exp(-1.0 / srate / parameters.attack)
    self.release = exp(-1.0 / srate / parameters.release)
    self.threshold = parameters.threshold
    self.factor = parameters.factor
    self.makeupGain = parameters.makeupGainValue()

    if let limit = parameters.clipLimit {
      let limitParams = LimiterParameters(clipLimit: limit, softClip: parameters.softClipValue())
      self.limiter = LimiterFilter(parameters: limitParams)
    } else {
      self.limiter = nil
    }
  }

  private func sumMonitorChannels(from chunk: AudioChunk) {
    let count = chunk.validFrames
    let ch0 = monitorChannels[0]
    guard let src0Base = chunk[ch0].baseAddress else { return }
    scratch.withUnsafeMutableBufferPointer { destBuf in
      guard let destBase = destBuf.baseAddress else { return }
      destBase.update(from: src0Base, count: count)
      for chIdx in 1..<monitorChannels.count {
        let ch = monitorChannels[chIdx]
        guard let srcBase = chunk[ch].baseAddress else { continue }
        vDSP_vaddD(destBase, 1, srcBase, 1, destBase, 1, vDSP_Length(count))
      }
    }
  }

  private func estimateLoudness(count: Int) {
    var prev = prevLoudness
    for i in 0..<count {
      var val = 20.0 * log10(abs(scratch[i]) + 1e-9)
      if val >= prev {
        val = attack * prev + (1.0 - attack) * val
      } else {
        val = release * prev + (1.0 - release) * val
      }
      prev = val
      scratch[i] = val
    }
    prevLoudness = prev
  }

  private func calculateLinearGain(count: Int) {
    for i in 0..<count {
      var val = scratch[i]
      if val > threshold {
        val = -(val - threshold) * (factor - 1.0) / factor
      } else {
        val = 0.0
      }
      val += makeupGain
      scratch[i] = PrcFmt.fromDB(val)
    }
  }

  private func applyGain(to waveform: MutableWaveform, count: Int) {
    guard let waveBase = waveform.baseAddress else { return }
    scratch.withUnsafeBufferPointer { scratchBuf in
      guard let scratchBase = scratchBuf.baseAddress else { return }
      vDSP_vmulD(waveBase, 1, scratchBase, 1, waveBase, 1, vDSP_Length(count))
    }
  }

  private func applyLimiter(to waveform: MutableWaveform, count: Int) {
    if let limiter = limiter {
      let sliced = MutableWaveform(start: waveform.baseAddress, count: count)
      limiter.process(waveform: sliced)
    }
  }

  func process(chunk: inout AudioChunk) throws {
    let count = chunk.validFrames
    guard count > 0 else { return }

    if scratch.count < count {
      scratch = [PrcFmt](repeating: 0.0, count: count)
    }

    sumMonitorChannels(from: chunk)
    estimateLoudness(count: count)
    calculateLinearGain(count: count)

    for ch in processChannels {
      let wave = chunk[ch]
      applyGain(to: wave, count: count)
      applyLimiter(to: wave, count: count)
    }
  }

  func updateParameters(_ config: ProcessorConfig, sampleRate: Int) {
    guard case .compressor(let p) = config else { return }

    var monitor = p.monitorChannelsArray()
    if monitor.isEmpty {
      monitor = Array(0..<p.channels)
    }
    self.monitorChannels = monitor

    var process = p.processChannelsArray()
    if process.isEmpty {
      process = Array(0..<p.channels)
    }
    self.processChannels = process

    let srate = PrcFmt(sampleRate)
    self.attack = exp(-1.0 / srate / p.attack)
    self.release = exp(-1.0 / srate / p.release)
    self.threshold = p.threshold
    self.factor = p.factor
    self.makeupGain = p.makeupGainValue()

    if let limit = p.clipLimit {
      let limitParams = LimiterParameters(clipLimit: limit, softClip: p.softClipValue())
      if let existingLimiter = self.limiter {
        existingLimiter.updateParameters(.limiter(limitParams), sampleRate: sampleRate)
      } else {
        self.limiter = LimiterFilter(parameters: limitParams)
      }
    } else {
      self.limiter = nil
    }
  }
}

File: CLib/Processors/compressor_processor.c
/**
 * @file compressor_processor.c
 * @brief Implementation of the dynamic range compressor processor.
 *
 * Implementation details:
 * - Exponential smoothing coefficients attack and release are precomputed as:
 *   attack = exp(-1.0 / (sample_rate * attack_time))
 *   release = exp(-1.0 / (sample_rate * release_time))
 * - Real-time processing (`compressor_processor_process`):
 *   1. Sums monitored channel waveforms into a pre-allocated scratch buffer using vDSP_vaddD (Apple Accelerate) or scalar addition.
 *   2. Envelope Detection: Computes instantaneous dB loudness and smooths it using attack filter when level rises, and release filter when level falls.
 *   3. Gain Reduction Curve: Applies compression ratio factor above threshold: -(val - threshold) * (factor - 1.0) / factor, adds makeup gain, and converts from dB to linear gain.
 *   4. Multiplies processed channels by linear gain curve using vDSP_vmulD (Apple Accelerate) or scalar multiplication.
 *   5. Optionally applies post-compression limiter filter.
 */

#include "compressor_processor.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>

#ifdef __APPLE__
#include <Accelerate/Accelerate.h>
#endif

compressor_processor_t* compressor_processor_create(const char* name, const compressor_parameters_t* params, int sample_rate, size_t chunk_size) {
    if (!params || sample_rate <= 0 || chunk_size == 0) return NULL;

    compressor_processor_t* processor = (compressor_processor_t*)calloc(1, sizeof(compressor_processor_t));
    if (!processor) return NULL;

    if (name) {
        strncpy(processor->name, name, sizeof(processor->name) - 1);
        processor->name[sizeof(processor->name) - 1] = '\0';
    } else {
        strcpy(processor->name, "compressor");
    }

    processor->scratch_capacity = chunk_size;
    processor->scratch = (prc_fmt_t*)calloc(chunk_size, sizeof(prc_fmt_t));
    if (!processor->scratch) {
        free(processor);
        return NULL;
    }

    if (params->monitor_channels_count > 0 && params->monitor_channels) {
        processor->monitor_channels_count = params->monitor_channels_count;
        processor->monitor_channels = (int*)calloc(processor->monitor_channels_count, sizeof(int));
        memcpy(processor->monitor_channels, params->monitor_channels, processor->monitor_channels_count * sizeof(int));
    } else {
        processor->monitor_channels_count = (size_t)params->channels;
        processor->monitor_channels = (int*)calloc(processor->monitor_channels_count, sizeof(int));
        for (size_t i = 0; i < processor->monitor_channels_count; i++) {
            processor->monitor_channels[i] = (int)i;
        }
    }

    if (params->process_channels_count > 0 && params->process_channels) {
        processor->process_channels_count = params->process_channels_count;
        processor->process_channels = (int*)calloc(processor->process_channels_count, sizeof(int));
        memcpy(processor->process_channels, params->process_channels, processor->process_channels_count * sizeof(int));
    } else {
        processor->process_channels_count = (size_t)params->channels;
        processor->process_channels = (int*)calloc(processor->process_channels_count, sizeof(int));
        for (size_t i = 0; i < processor->process_channels_count; i++) {
            processor->process_channels[i] = (int)i;
        }
    }

    if (!processor->monitor_channels || !processor->process_channels) {
        compressor_processor_free(processor);
        return NULL;
    }

    prc_fmt_t srate = (prc_fmt_t)sample_rate;
    processor->attack = exp(-1.0 / srate / params->attack);
    processor->release = exp(-1.0 / srate / params->release);
    processor->threshold = params->threshold;
    processor->factor = params->factor;
    processor->makeup_gain = params->has_makeup_gain ? params->makeup_gain : 0.0;
    processor->prev_loudness = -100.0;

    if (params->has_clip_limit) {
        limiter_parameters_t limit_params = {0};
        limit_params.clip_limit = params->clip_limit;
        limit_params.soft_clip = params->soft_clip;
        processor->limiter = limiter_filter_create("limiter", &limit_params);
    } else {
        processor->limiter = NULL;
    }

    return processor;
}

void compressor_processor_free(compressor_processor_t* processor) {
    if (!processor) return;
    free(processor->monitor_channels);
    free(processor->process_channels);
    free(processor->scratch);
    if (processor->limiter) limiter_filter_free(processor->limiter);
    free(processor);
}

void compressor_processor_process(compressor_processor_t* processor, audio_chunk_t* chunk) {
    if (!processor || !chunk || !processor->scratch) return;
    size_t count = chunk->valid_frames;
    if (count > processor->scratch_capacity) count = processor->scratch_capacity;
    if (count == 0 || processor->monitor_channels_count == 0) return;

    // Step 1: Sum monitored channels into scratch buffer to evaluate overall signal level
    int ch0 = processor->monitor_channels[0];
    const double* src0_base = audio_chunk_get_channel(chunk, ch0);
    if (!src0_base) return;
    memcpy(processor->scratch, src0_base, count * sizeof(double));

    for (size_t ch_idx = 1; ch_idx < processor->monitor_channels_count; ch_idx++) {
        int ch = processor->monitor_channels[ch_idx];
        const double* src_base = audio_chunk_get_channel(chunk, ch);
        if (!src_base) continue;
#ifdef __APPLE__
        vDSP_vaddD(processor->scratch, 1, src_base, 1, processor->scratch, 1, count);
#else
        for (size_t i = 0; i < count; i++) {
            processor->scratch[i] += src_base[i];
        }
#endif
    }

    // Step 2: Envelope Detection (Loudness Estimation with Attack/Release Smoothing)
    double prev = processor->prev_loudness;
    for (size_t i = 0; i < count; i++) {
        double val = 20.0 * log10(fabs(processor->scratch[i]) + 1e-9);
        if (val >= prev) {
            // Signal level rising: apply attack time constant
            val = processor->attack * prev + (1.0 - processor->attack) * val;
        } else {
            // Signal level falling: apply release time constant
            val = processor->release * prev + (1.0 - processor->release) * val;
        }
        prev = val;
        processor->scratch[i] = val;
    }
    processor->prev_loudness = prev;

    // Step 3: Gain Reduction Curve Calculation
    for (size_t i = 0; i < count; i++) {
        double val = processor->scratch[i];
        if (val > processor->threshold) {
            // Above threshold: attenuate according to compression ratio (factor)
            val = -(val - processor->threshold) * (processor->factor - 1.0) / processor->factor;
        } else {
            // Below threshold: unity gain (0.0 dB reduction)
            val = 0.0;
        }
        val += processor->makeup_gain;
        // Convert gain reduction in dB to linear gain multiplier
        processor->scratch[i] = prc_fmt_from_db(val);
    }

    // Step 4: Apply linear gain to all processed channels
    for (size_t ch_idx = 0; ch_idx < processor->process_channels_count; ch_idx++) {
        int ch = processor->process_channels[ch_idx];
        double* wave = audio_chunk_get_channel(chunk, ch);
        if (!wave) continue;
#ifdef __APPLE__
        vDSP_vmulD(wave, 1, processor->scratch, 1, wave, 1, count);
#else
        for (size_t i = 0; i < count; i++) {
            wave[i] *= processor->scratch[i];
        }
#endif
        // Step 5: Optionally run post-compression limiter to prevent clipping
        if (processor->limiter) {
            limiter_filter_process(processor->limiter, wave, count);
        }
    }
}


void compressor_processor_update_parameters(compressor_processor_t* processor, const processor_config_t* config, int sample_rate) {
    if (!processor || !config || sample_rate <= 0) return;
    if (config->type != PROCESSOR_TYPE_COMPRESSOR) return;
    const compressor_parameters_t* params = &config->parameters.compressor;

    if (params->monitor_channels_count > 0 && params->monitor_channels) {
        free(processor->monitor_channels);
        processor->monitor_channels_count = params->monitor_channels_count;
        processor->monitor_channels = (int*)calloc(processor->monitor_channels_count, sizeof(int));
        memcpy(processor->monitor_channels, params->monitor_channels, processor->monitor_channels_count * sizeof(int));
    }
    if (params->process_channels_count > 0 && params->process_channels) {
        free(processor->process_channels);
        processor->process_channels_count = params->process_channels_count;
        processor->process_channels = (int*)calloc(processor->process_channels_count, sizeof(int));
        memcpy(processor->process_channels, params->process_channels, processor->process_channels_count * sizeof(int));
    }

    prc_fmt_t srate = (prc_fmt_t)sample_rate;
    processor->attack = exp(-1.0 / srate / params->attack);
    processor->release = exp(-1.0 / srate / params->release);
    processor->threshold = params->threshold;
    processor->factor = params->factor;
    processor->makeup_gain = params->has_makeup_gain ? params->makeup_gain : 0.0;

    if (params->has_clip_limit) {
        limiter_parameters_t limit_params = {0};
        limit_params.clip_limit = params->clip_limit;
        limit_params.soft_clip = params->soft_clip;
        if (processor->limiter) {
            filter_config_t fconfig = {0};
            fconfig.type = FILTER_TYPE_LIMITER;
            fconfig.parameters.limiter = limit_params;
            limiter_filter_update_parameters(processor->limiter, &fconfig, sample_rate);
        } else {
            processor->limiter = limiter_filter_create("limiter", &limit_params);
        }
    } else {
        if (processor->limiter) {
            limiter_filter_free(processor->limiter);
            processor->limiter = NULL;
        }
    }
}
